// motion/lib/helpers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRandomNumber = getRandomNumber;
exports.getNpmErrorMessage = getNpmErrorMessage;
function getRandomNumber(min, max) {
  return Math.floor(Math.random() * (max - min)) + min;
}

var TICK = exports.TICK = '✓';
var CONFIG_FILE_NAME = exports.CONFIG_FILE_NAME = '.motion.json';
var CONFIG_FILE_DEFAULT = exports.CONFIG_FILE_DEFAULT = {
  babel: {
    plugins: [],
    presets: ['babel-preset-steelbrain']
  },
  pundle: {
    presets: [],
    components: []
  },
  pathType: 'filePath',
  saveNpmModules: true,
  get webServerPort() {
    return getRandomNumber(8000, 15000);
  },
  outputDirectory: './dist'
};
var CONFIG_FILE_OPTIONS = exports.CONFIG_FILE_OPTIONS = {
  prettyPrint: true,
  createIfNonExistent: false
};

var NPM_ERROR_EXTRACTION_REGEX = /npm ERR! code .*?\n\nnpm ERR! .*? (.*)/gm;
function getNpmErrorMessage(contents) {
  if (!contents.includes('npm ERR!')) {
    return contents;
  }
  var match = NPM_ERROR_EXTRACTION_REGEX.exec(contents);
  if (match) {
    return match[1];
  }
  return contents;
}
