// pundle-api/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FILE_FEATURES = exports.Context = exports.FileChunk = exports.File = exports.version = undefined;

var _helpers = require('./helpers');

Object.defineProperty(exports, 'version', {
  enumerable: true,
  get: function get() {
    return _helpers.version;
  }
});

var _components = require('./components');

Object.keys(_components).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _components[key];
    }
  });
});

var _issues = require('./issues');

Object.keys(_issues).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _issues[key];
    }
  });
});

var _rules = require('./rules');

Object.keys(_rules).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _rules[key];
    }
  });
});
exports.getRelativeFilePath = getRelativeFilePath;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _context = require('./context');

var _context2 = _interopRequireDefault(_context);

var _file = require('./file');

var _fileChunk = require('./file-chunk');

var _fileChunk2 = _interopRequireDefault(_fileChunk);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getRelativeFilePath(filePath, rootDirectory) {
  var relative = _path2.default.relative(rootDirectory, filePath);
  if (relative.slice(0, 2) === '..') {
    return filePath;
  }
  return relative;
}

exports.File = _file.File;
exports.FileChunk = _fileChunk2.default;
exports.Context = _context2.default;
exports.FILE_FEATURES = _file.FEATURES;
